File: pa4/List.c
//-----------------------------------------------------------------------------
//  List.c
//  Name: Eric Mar
//  CruzID: emmar
//
//  List ADT with get functions, comparison function, list modification function
//  List transversal function, list copy function, and print function.
//-----------------------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h> 
#include <string.h>
#include "List.h" // Include user-defined header file

/*
 * Node Structure
 */
typedef struct NodeObj{
   int data;
   struct NodeObj* prev;
   struct NodeObj* next;
} NodeObj;

/*
 * List Structure
 */
typedef struct ListObj{
   struct NodeObj* cursor;
   struct NodeObj* front;
   struct NodeObj* back;
   int index;
   int length;
} ListObj;

// Constructors-Destructors ---------------------------------------------------
List newList(void) {
	List L = malloc (sizeof(ListObj));
	L->cursor = NULL;
	L->front = NULL;
	L->back = NULL;
	L->index = -1;
	L->length = 0;
	return L;
}

void freeList(List* pL) {
	if (pL == NULL){
		fprintf(stderr, "List Error: freeList() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	if (*pL == NULL){
		return;
	}

	List L = *pL;

	if (L->front == NULL)
		return;

	Node ptr = L->front;
	Node temp = ptr;
	while (L->length != 0) {
		ptr = ptr->next;
		free(temp);
		temp = ptr;
		L->length--;
	}
	L->cursor = NULL;
	L->front = NULL;
	L->back = NULL;
	L->index = -1;
	L->length = 0;
	*pL = NULL;
}

// Access functions -----------------------------------------------------------
int length(List L) {
	if (L == NULL){
		fprintf(stderr, "List Error: length() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	return L->length;
}

int index(List L) {
	if (L == NULL){
		fprintf(stderr, "List Error: index() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	return L->index;
}

int front(List L) {
	if (L == NULL){
		fprintf(stderr, "List Error: front() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	return L->front->data;
}

int back(List L) {
	if (L == NULL){
		fprintf(stderr, "List Error: back() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	return L->back->data;
}

int get(List L) {
	if (L == NULL){
		fprintf(stderr, "List Error: get() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	return L->cursor->data;
}

int equals(List A, List B) {
	if (A == NULL){
		fprintf(stderr, "List Error: equals() called on NULL List A reference\n");
		exit(EXIT_FAILURE);
	}

	if (B == NULL){
		fprintf(stderr, "List Error: equals() called on NULL List B reference\n");
		exit(EXIT_FAILURE);
	}

	Node p = A->front;
	Node q = B->front;

	if(A->length != B->length)
		return 0;
	for(int i = 0; i < A->length; i++){
		if (p->data != q->data)
			return 0;
		p = p->next;
		q = q->next;
	}
	return 1;
}

// Manipulation procedures ----------------------------------------------------
void clear(List L) {
	if (L == NULL){
		fprintf(stderr, "List Error: clear() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	freeList(&L);
}

void moveFront(List L) {
	if (L == NULL){
		fprintf(stderr, "List Error: moveFront() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	if (L->length != 0) {
		L->cursor = L->front;
		L->index = 0;
	}
}

void moveBack(List L) {
	if (L == NULL){
		fprintf(stderr, "List Error: moveBack() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	if (L->length != 0) {
		L->cursor = L->back;
		L->index = L->length - 1;
	}
}

void movePrev(List L) {
	if (L == NULL){
		fprintf(stderr, "List Error: movePrev() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	if (L->cursor != NULL){
		if (L->cursor == L->front){
			L->cursor = NULL;
			L->index = -1;
		}
		else{
			L->cursor = L->cursor->prev;
			L->index--;
		}
	}
}

void moveNext(List L) {
	if (L == NULL){
		fprintf(stderr, "List Error: moveNext() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	if (L->cursor != NULL){
		if (L->cursor == L->back){
			L->cursor = NULL;
			L->index = -1;
		}
		else{
			L->cursor = L->cursor->next;
			L->index++;
		}
	}
}

void prepend(List L, int data) {
	if (L == NULL){
		fprintf(stderr, "List Error: prepend() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	Node N = malloc (sizeof(NodeObj));
	N->data = data;
	N->next = L->front;
	if (L->front != NULL)
		L->front->prev = N;
	if (L->back == NULL)
		L->back = N;
	L->front = N;
	L->length++;
	if (L->cursor != NULL)
		L->index++;
}

void append(List L, int data) {
	if (L == NULL){
		fprintf(stderr, "List Error: append() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	Node N = malloc (sizeof(NodeObj));
	N->data = data;
	N->prev = L->back;
	if (L->back != NULL)
		L->back->next = N;
	if (L->length == 0)
		L->front = N;
	L->back = N;
	L->length++;
}

void insertBefore(List L, int data) {
	if (L == NULL){
		fprintf(stderr, "List Error: insertBefore() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	if (!(L->length > 0))
		fprintf(stderr, "No Elements in List");
	else if (!(L->index >= 0))
		fprintf(stderr, "No Cursor Identified");
	else {
		if (L->cursor == L->front)
			prepend(L, data);
		else {
			Node N = malloc (sizeof(NodeObj));
			N->data = data;
			N->next = L->cursor;
			N->prev = L->cursor->prev;
			L->cursor->prev->next = N;
			L->cursor->prev = N;
			L->length++;
			L->index++;
		}
	}
}

void insertAfter(List L, int data) {
	if (L == NULL){
		fprintf(stderr, "List Error: insertAfter() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	if (!(L->length > 0))
		fprintf(stderr, "No Elements in List");
	else if (!(L->index >= 0))
		fprintf(stderr, "No Cursor Identified");
	else {
		if (L->cursor == L->back)
			append(L, data);
		else {
			Node N = malloc (sizeof(NodeObj));
			N->data = data;
			N->next = L->cursor->next;
			N->prev = L->cursor;
			L->cursor->next->prev = N;
			L->cursor->next = N;
			L->length++;
		}
	}
}

void deleteFront(List L) {
	if (L == NULL){
		fprintf(stderr, "List Error: deleteFront() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	if (!(L->length > 0))
		fprintf(stderr, "No Elements in List");
	else{
		if (L->cursor == L->front){
			L->cursor = NULL;
			L->index = -1;
		}
		if (L->back == L->front)
			L->back = NULL;
		L->front = L->front->next;
		if (L->index >= 0)
			L->index--;
		L->length--;
	}
}

void deleteBack(List L) {
	if (L == NULL){
		fprintf(stderr, "List Error: deleteBack() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	if (!(L->length > 0))
		fprintf(stderr, "No Elements in List");
	else{
		if (L->cursor == L->back){
			L->cursor = NULL;
			L->index = -1;
		}
		if (L->front == L->back)
			L->front = NULL;
		L->back = L->back->prev;
		L->length--;
	}
}

void delete(List L) {
	if (L == NULL){
		fprintf(stderr, "List Error: delete() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	if (!(L->length > 0))
		fprintf(stderr, "No Elements in List");
	else if (!(L->index >= 0))
		fprintf(stderr, "No Cursor Identified");
	else {
		if (L->cursor == L->front)
			deleteFront(L);
		else if (L->cursor == L->back)
			deleteBack(L);
		else{
			L->cursor->prev = L->cursor->next;
			L->cursor->next->prev = L->cursor->prev;
			L->cursor = NULL;
			L->index = -1;
			L->length--;
		}
	}
}

// Other operations -----------------------------------------------------------
void printList(FILE* out, List L) {
	if (L == NULL){
		fprintf(stderr, "List Error: printList() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	if (out == NULL){
		fprintf(stderr, "List Error: printList() called on NULL out reference\n");
		exit(EXIT_FAILURE);
	}

	if (L->front != NULL){
		fprintf(out, "%d", L->front->data);
		Node temp = L->front->next;
		while (temp != NULL){
     		fprintf(out, " %d", temp->data);
			temp = temp->next;
		}
	}
}

List copyList(List L) {
	if (L == NULL){
		fprintf(stderr, "List Error: copyList() called on NULL List reference\n");
		exit(EXIT_FAILURE);
	}

	List clone = newList();
	Node temp = L->front;
	for(int i = 0; i < L->length; i++){
		append(clone, temp->data);
		temp = temp->next;
	}
	return clone;
}

File: pa3/List.java
//-----------------------------------------------------------------------------
//  List.java
//  Name: Eric Mar
//  CruzID: emmar
//
//  List ADT with get functions, comparison function, list modification function
//  List transversal function, list copy function, and print function.
//-----------------------------------------------------------------------------

public class List {

	private Node cursor;
	private Node front;
	private Node back;
	private int index;
	private int length;

	// Constructor
	// Creates a new empty list.
	List() {
		cursor = null;
		front = null;
		back = null;
		index = -1;
		length = 0;
	}

	// Access functions

	// Returns the number of elements in this List.
	int length() {
		return length;
	}

	// If cursor is defined, returns the index of the cursor element,
	// otherwise returns -1.
	int index() {
		return index;
	}

	// Returns front element. Pre: length()>0
	Object front() {
		if (!(length > 0)){
			System.err.println("No Elements in List");
			return null;
		}
		return front.data;
	}

	// Returns back element. Pre: length()>0
	Object back() {
		if (!(length > 0)){
			System.err.println("No Elements in List");
			return null;
		}
		return back.data;
	}

	// Returns cursor element. Pre: length()>0, index()>=0
	Object get() {
		if (!(length > 0)){
			System.err.println("No Elements in List");
			return null;
		}
		return cursor.data;
	}

	// Returns true if and only if this List and L are the same
	// integer sequence. The states of the cursors in the two Lists
	// are not used in determining equality.
	public boolean equals(Object X) {
		if (X instanceof List){
			List L = (List)X;

			Node p = this.front;
			Node q = L.front;

			
			if (this.length() != L.length())
				return false;
			while(p != null && q != null){
				if (p.data != q.data)
					return false;
				p = p.next;
				q = q.next;
			}
			return true;
		}

		return false;
	}

	// Manipulation procedures

	// Resets this List to its original empty state.
	void clear() {
		this.cursor = null;
		this.front = null;
		this.back = null;
		this.index = -1;
		this.length = 0;
	}

	// If List is non-empty, places the cursor under the front element,
	// otherwise does nothing.
	void moveFront() {
		if (length != 0) {
			cursor = front;
			index = 0;
		}
	}

	// If List is non-empty, places the cursor under the back element,
	// otherwise does nothing.
	void moveBack() {
		if (length != 0) {
			cursor = back;
			index = length - 1;
		}
	}

	// If cursor is defined and not at front, moves cursor one step toward
	// front of this List, if cursor is defined and at front, cursor becomes
	// undefined, if cursor is undefined does nothing.
	void movePrev() {
		if (cursor != null){
			if (cursor == front){
				cursor = null;
				index = -1;
			}
			else{
				cursor = cursor.prev;
				index--;
			}
		}
	}

	// If cursor is defined and not at back, moves cursor one step toward
	// back of this List, if cursor is defined and at back, cursor becomes
	// undefined, if cursor is undefined does nothing.
	void moveNext() {
		if (cursor != null){
			if (cursor == back){
				cursor = null;
				index = -1;
			}
			else{
				cursor = cursor.next;
				index++;
			}
		}
	}

	// Insert new element into this List. If List is non-empty,
	// insertion takes place before front element.
	void prepend(Object data) {
		Node insert = new Node (data, front, null);
		if (front != null)
			this.front.prev = insert;
		if (back == null)
			this.back = insert;
		this.front = insert;
		this.length++;
		if (cursor != null)
			index++;
	}

	// Insert new element into this List. If List is non-empty,
	// insertion takes place after back element.
	void append(Object data) {
		Node insert = new Node (data, null, back);
		if (back != null)
			this.back.next = insert;
		if (front == null)
			this.front = insert;
		this.back = insert;
		this.length++;
	}

	// Insert new element before cursor.
	// Pre: length()>0, index()>=0
	void insertBefore(Object data) {
		if (!(length > 0))
			System.err.println("No Elements in List");
		else if (!(index >= 0))
			System.err.println("No Cursor Identified");
		else {
			if (cursor == front)
				prepend(data);
			else {
				Node insert = new Node (data, cursor, cursor.prev);
				this.cursor.prev.next = insert;
				this.cursor.prev = insert;
				this.length++;
				index++;
			}
		}
	}

	// Inserts new element after cursor.
	// Pre: length()>0, index()>=0
	void insertAfter(Object data) {
		if (!(length > 0))
			System.err.println("No Elements in List");
		else if (!(index >= 0))
			System.err.println("No Cursor Identified");
		else {
			if (cursor == back)
				append(data);
			else {
				Node insert = new Node (data, cursor.next, cursor);
				this.cursor.next.prev = insert;
				this.cursor.next = insert;
				this.length++;
			}
		}
	}

	// Deletes the front element. Pre: length()>0
	void deleteFront() {
		if (!(length > 0))
			System.err.println("No Elements in List");
		else{
			if (cursor == front){
				cursor = null;
				index = -1;
			}
			if (back == front)
				this.back = null;
			this.front = front.next;
			if (index >= 0)
				index--;
			length--;
		}
	}

	// Deletes the back element. Pre: length()>0
	void deleteBack() {
		if (!(length > 0))
			System.err.println("No Elements in List");
		else{
			if (cursor == back){
				cursor = null;
				index = -1;
			}
			if (front == back)
				this.front = null;
			this.back = back.prev;
			length--;
		}
	}

	// Deletes cursor element, making cursor undefined.
	// Pre: length()>0, index()>=0

	void delete() {
		if (!(length > 0))
			System.err.println("No Elements in List");
		else if (!(index >= 0))
			System.err.println("No Cursor Identified");
		else{
			if (cursor == front)
				deleteFront();
			else if (cursor == back)
				deleteBack();
			else{
				cursor.prev = cursor.next;
				cursor.next.prev = cursor.prev;
				cursor = null;
				index = -1;
				length--;
			}
		}
	}

	// Other methods

	// Overrides Object's toString method. Returns a String
	// representation of this List consisting of a space
	// separated sequence of integers, with front on left.
	public String toString() {
		if (front != null){
			String s = front.toString();
			Node temp = front.next;
			while (temp != null){
				s += " " + temp.toString();
				temp = temp.next;
			}
			return s;
		}
		return "";
	}

	private class Node {
		private Object data;
		private Node next;
		private Node prev;

		Node(Object d, Node n, Node p){
			data = d;
			next = n;
			prev = p;
		}

		public String toString() {
			return "" + data;
		}
	}
}